pkg/pool-utils/contracts/rates/PriceRateCache.sol

```diff
@@ -23,55 +23,71 @@ import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
-    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 128 + 64;
+    uint256 private constant _CURRENT_PRICE_RATE_OFFSET = 0;
+    uint256 private constant _POST_JOINEXIT_RATE_OFFSET = 96;
+    uint256 private constant _PRICE_RATE_CACHE_DURATION_OFFSET = 192;
+    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 224;
 
     /**
-     * @dev Returns the rate of a price rate cache.
+     * @dev Returns the current rate of a price rate cache.
      */
     function getRate(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_PRICE_RATE_CACHE_VALUE_OFFSET, 128);
+        return cache.decodeUint(_CURRENT_PRICE_RATE_OFFSET, 96);
+    }
+
+    /**
+     * @dev Returns the current rate of a price rate cache.
+     */
+    function getPostJoinExitRate(bytes32 cache) internal pure returns (uint256) {
+        return cache.decodeUint(_POST_JOINEXIT_RATE_OFFSET, 96);
+    }
+
+    /**
+     * @dev Sets the post join/exit rate in an existing cache record, and returns it.
+     */
+    function setPostJoinExitRate(bytes32 cache, uint256 rate) internal pure returns (bytes32) {
+        _require(rate >> 96 == 0, Errors.PRICE_RATE_OVERFLOW);
+
+        return cache.insertUint(rate, _POST_JOINEXIT_RATE_OFFSET, 96);
     }
 
     /**
      * @dev Returns the duration of a price rate cache.
      */
     function getDuration(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, 64);
+        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, 32);
     }
 
     /**
      * @dev Returns the duration and expiration time of a price rate cache.
      */
     function getTimestamps(bytes32 cache) internal pure returns (uint256 duration, uint256 expires) {
         duration = getDuration(cache);
-        expires = cache.decodeUint(_PRICE_RATE_CACHE_EXPIRES_OFFSET, 64);
+        expires = cache.decodeUint(_PRICE_RATE_CACHE_EXPIRES_OFFSET, 32);
     }
 
     /**
      * @dev Encodes rate and duration into a price rate cache. The expiration time is computed automatically, counting
      * from the current time.
      */
     function encode(uint256 rate, uint256 duration) internal view returns (bytes32) {
-        _require(rate >> 128 == 0, Errors.PRICE_RATE_OVERFLOW);
+        _require(rate >> 96 == 0, Errors.PRICE_RATE_OVERFLOW);
 
         // solhint-disable not-rely-on-time
         return
-            WordCodec.encodeUint(rate, _PRICE_RATE_CACHE_VALUE_OFFSET, 128) |
-            WordCodec.encodeUint(duration, _PRICE_RATE_CACHE_DURATION_OFFSET, 64) |
-            WordCodec.encodeUint(block.timestamp + duration, _PRICE_RATE_CACHE_EXPIRES_OFFSET, 64);
+            WordCodec.encodeUint(rate, _CURRENT_PRICE_RATE_OFFSET, 96) |
+            WordCodec.encodeUint(duration, _PRICE_RATE_CACHE_DURATION_OFFSET, 32) |
+            WordCodec.encodeUint(block.timestamp + duration, _PRICE_RATE_CACHE_EXPIRES_OFFSET, 32);
```

Comment:
This would clear the 'old rate' field. Maybe we want to change this to `updateRate`, and instead of encoding these three we instead insert the expiration (now + duration) and the new rate.